System/Directory/Internal/Common/AbstractFilePath.hs

```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE CPP #-}
+
+module System.Directory.Internal.Common.AbstractFilePath where
+
+import System.AbstractFilePath (unpackAFP, packAFP, unsafeFromChar)
+import qualified System.AbstractFilePath as AFP
+
+import qualified System.File.AbstractFilePath as P
+
+import System.OsString.Internal.Types
+import System.AbstractFilePath.Types
+import System.AbstractFilePath.Data.ByteString.Short.Decode
+
+#define SYSTEM_FILEPATH_MODULE System.AbstractFilePath
+#define FILEPATH AbstractFilePath
+#include "Template.hs"
+
+unpack :: AbstractFilePath -> [OsChar]
+unpack = unpackAFP
+
+pack :: [OsChar] -> AbstractFilePath
+pack = packAFP
+
+-- UNSAFE... only use this with ascii, not unknown input
+unsafeFromChar' :: Char -> OsChar
+unsafeFromChar' = unsafeFromChar
```

Comment:
Indeed unsafe, due to truncation. But we make sure we only use it on ascii input.